pkg/scheduler/scheduler.go

```diff
@@ -56,43 +62,68 @@ func getNodeList(topo *csi.TopologyRequirement) ([]string, error) {
 
 // runScheduler goes through the node mapping
 // in the topology and picks the node which is less weighted
-func runScheduler(nodelist []string, nmap map[string]int64) string {
-	var selected string
+func runScheduler(nodelist []string, nmap map[string]int64) []string {
+	var preferred []string
+	var fmap []kv
 
-	var weight int64 = math.MaxInt64
-
-	// schedule it on the node which has less weight
+	// go though the filtered node and prepare the preferred list
 	for _, node := range nodelist {
-		if nmap[node] < weight {
-			selected = node
-			weight = nmap[node]
+		if val, ok := nmap[node]; ok {
+			// create the filtered node map
+			fmap = append(fmap, kv{node, val})
+		} else {
+			// put the non occupied nodes in beginning of the list
+			preferred = append(preferred, node)
 		}
 	}
-	return selected
+
+	// sort the filtered node map
+	sort.Slice(fmap, func(i, j int) bool {
+		return fmap[i].Value < fmap[j].Value
+	})
+
+	// put the occupied nodes in the sorted order at the end
+	for _, kv := range fmap {
+		preferred = append(preferred, kv.Key)
+	}
+
+	return preferred
 }
 
 // Scheduler schedules the PV as per topology constraints for
 // the given node weight.
-func Scheduler(req *csi.CreateVolumeRequest, nmap map[string]int64) string {
-	topo := req.AccessibilityRequirements
-	if topo == nil ||
-		len(topo.Preferred) == 0 {
+func Scheduler(req *csi.CreateVolumeRequest, nmap map[string]int64) []string {
```

Comment:
q_: is it required to maintain backward compatibility for this API so that any of the CSI drivers using this method can still work, after this is merged and if they are using the latest code from this repo (without changing their code.

Reply:
it will work even after the merge because of the gomod.